drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+	uint8_t rx_buf[3] = {0x00};
+	uint8_t chip_id[3] = {0x00};
+	int ret = 0;
+	/* Read chip ID (can only be read in sleep mode)*/
+	if (bmm350_reg_read(dev, BMM350_REG_CHIP_ID, &chip_id[0], 3) < 0) {
+		LOG_ERR("failed reading chip id");
+		goto err_poweroff;
+	}
+	if (chip_id[2] != BMM350_CHIP_ID) {
+		LOG_ERR("invalid chip id 0x%x", chip_id[2]);
+		goto err_poweroff;
+	}
+	/* Soft-reset */
+	soft_reset = BMM350_CMD_SOFTRESET;
+
+	/* Set the command in the command register */
+	ret = bmm350_reg_write(dev, BMM350_REG_CMD, soft_reset);
+	k_usleep(BMM350_SOFT_RESET_DELAY);
+	/* Read chip ID (can only be read in sleep mode)*/
+	if (bmm350_reg_read(dev, BMM350_REG_CHIP_ID, &chip_id[0], 3) < 0) {
+		LOG_ERR("failed reading chip id");
+		goto err_poweroff;
+	}
+	if (chip_id[2] != BMM350_CHIP_ID) {
+		LOG_ERR("invalid chip id 0x%x", chip_id[2]);
+		goto err_poweroff;
+	}
+	ret = bmm350_otp_dump_after_boot(dev);
+	LOG_DBG("bmm350 chip_id 0x%x otp dump after boot %d\n", chip_id[2], ret);
+
+	if (bmm350_reg_write(dev, BMM350_REG_OTP_CMD_REG, BMM350_OTP_CMD_PWR_OFF_OTP) < 0) {
+		LOG_ERR("failed to set REP");
+		goto err_poweroff;
+	}
+
+	ret += bmm350_magnetic_reset_and_wait(dev);
+
+	LOG_DBG("bmm350 setup result %d\n", ret);
+
+	ret += bmm350_get_pmu_cmd_status_0(&pmu_cmd_stat_0, dev);
+	ret += bmm350_reg_read(dev, BMM350_REG_ERR_REG, &rx_buf[0], 3);
+	if (ret != 0) {
+		LOG_ERR("%s %d", __func__, ret);
+	}
+
+	return 0;
+
+err_poweroff:
+	ret = bmm350_set_powermode(BMM350_SUSPEND_MODE, dev);
+	if (ret != 0) {
+		return -EIO;
+	}
+	return -EIO;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		ret = bmm350_set_powermode(BMM350_NORMAL_MODE, dev);
+		if (ret != 0) {
+			LOG_ERR("failed to enter normal mode: %d", ret);
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = bmm350_set_powermode(BMM350_SUSPEND_MODE, dev);
+		if (ret != 0) {
+			LOG_ERR("failed to enter suspend mode: %d", ret);
+		}
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return ret;
+}
+#endif
+
+static int bmm350_init(const struct device *dev)
+{
+	int err = 0;
+	struct bmm350_data *data = dev->data;
+
+	err = bmm350_bus_check(dev);
+	if (err < 0) {
+		LOG_ERR("bus check failed: %d", err);
+		return err;
+	}
+
+	if (bmm350_init_chip(dev) < 0) {
+		LOG_ERR("failed to initialize chip");
+		return -EIO;
+	}
+
+#ifdef CONFIG_BMM350_TRIGGER
+	if (bmm350_trigger_mode_init(dev) < 0) {
+		LOG_ERR("Cannot set up trigger mode.");
+		return -EINVAL;
+	}
+#endif
+
+	/* Assign axis_en with all axis enabled (BMM350_EN_XYZ_MSK) */
+	data->axis_en = BMM350_EN_XYZ_MSK;
+
+	return 0;
+}
+
+/* Initializes a struct bmm350_config for an instance on an I2C bus. */
+#define BMM350_CONFIG_I2C(inst) .bus.i2c = I2C_DT_SPEC_INST_GET(inst), .bus_io = &bmm350_bus_io_i2c,
+#ifdef CONFIG_BMM350_TRIGGER
+#define BMM350_INT_CFG(inst) .drdy_int = GPIO_DT_SPEC_INST_GET(inst, drdy_gpios),
```

Comment:
This will get you, if you have 2 BMM350s on your board one with a trigger and the other without then the Kconfig will be there but one node will not have `drdy_gpios`, so you can't switch on the Kconfig, use
```c
GPIO_DT_SPEC_INST_GET_OR(inst, drdy_gpios, { 0 })
```